src/main.c

```diff
@@ -3128,6 +3089,59 @@ static void setup_configurators(void)
     h2o_config_register_status_handler(&conf.globalconf, &extra_status_handler);
 }
 
+static void create_per_thread_listeners(void)
+{
+    for (size_t i = 0; i != conf.num_listeners; ++i) {
+        struct listener_config_t *listener_config = conf.listeners[i];
+            int reuseport = 0;
+
+#if H2O_USE_REUSEPORT
+            socklen_t reuseportlen = sizeof(reuseport);
+            if (getsockopt(listener_config->fds.entries[0], SOL_SOCKET, H2O_SO_REUSEPORT, &reuseport, &reuseportlen) != 0) {
+                perror("gestockopt(SO_REUSEPORT) failed");
+                abort();
+            }
+            assert(reuseportlen == sizeof(reuseport));
+#endif
+
+        for (size_t j = 1; j < conf.thread_map.size; j++) {
+            int fd = -1;
+#if H2O_USE_REUSEPORT
+            if (reuseport) {
+                int type;
+                socklen_t typelen = sizeof(type);
+                struct sockaddr_storage ss;
+                socklen_t sslen = sizeof(ss);
+                if (getsockopt(listener_config->fds.entries[0], SOL_SOCKET, SO_TYPE, &type, &typelen) != 0) {
+                    perror("failed to obtain the type of a listening socket");
+                    abort();
+                }
+                assert(type == SOCK_DGRAM || type == SOCK_STREAM);
+                if (getsockname(listener_config->fds.entries[0], (struct sockaddr *)&ss, &sslen) != 0) {
+                    perror("failed to obtain local address of a listening socket");
+                    abort();
+                }
+                if ((fd = open_listener(ss.ss_family, type, type == SOCK_STREAM ? IPPROTO_TCP : IPPROTO_UDP, (struct sockaddr *)&ss, sslen)) != -1) {
+                    if (type == SOCK_DGRAM)
+                        setsockopt_recvpktinfo(fd, ss.ss_family);
+                } else {
+                    reuseport = 0;
+                }
+            }
+#endif
+            if (!reuseport && (fd = dup(listener_config->fds.entries[0])) == -1) {
+                perror("failed to dup listening socket");
+                abort();
+            }
```

Comment:
I wonder if we could improve the maintenability of the code, by extracting the logic as our own super-dup function (maybe we can call it `dup_listener`); i.e. a function that accepts a file descriptor as an argument and returns a copy, with the copy being:
* the result of `dup` when `H2O_USE_REUSEPORT` is not set, or
* the result of `open_listener` that binds to the same address / port as the given file descriptor, when `H2O_USE_REUSEPORT` is set